File: frontend/src/types/conversation.ts
import type {
  AskAction,
  AskEvidenceItem,
  AskImpact,
  AskRepositoryCandidate,
} from "./ask";
import type { ProvenanceKind } from "../components/intelligence/ProvenanceTag";

export type ConversationMode = "general" | "migration" | "refinement";

export interface ConversationEntityRef {
  ref: string;
  name: string;
  impact_level: "low" | "medium" | "high" | null;
}

export interface MigrationRisk {
  label: string;
  reason: string;
  provenance: AskEvidenceItem["provenance"];
}

export interface MigrationScope {
  source_technology: string;
  target_technology: string;
  direct: string[];
  indirect: string[];
  risks: MigrationRisk[];
  primary_repository_id: string | null;
}

export type WorkItemType = "epic" | "story" | "task" | "spike";
export type WorkItemStatus = "existing" | "proposed";
export type EdgeRelationship = "blocks" | "depends_on" | "enables" | "related" | "parent_child";
export type QuestionCategory = "known" | "derived" | "assumption" | "unknown";
export type ReadinessLevel = "ready" | "mostly_ready" | "needs_clarification" | "not_ready";

export interface WorkItem {
  id: string;
  type: WorkItemType;
  status: WorkItemStatus;
  title: string;
  objective: string;
  context: string;
  scope: string;
  acceptance_criteria: string[];
  related_systems: string[];
  risks: string[];
  evidence_note: string;
  provenance: ProvenanceKind;
}

export interface WorkItemEdge {
  source_id: string;
  target_id: string;
  relationship: EdgeRelationship;
  provenance: ProvenanceKind;
  source_system: "jira" | "refinement_analysis";
}

export interface OpenQuestion {
  question: string;
  category: QuestionCategory;
  note: string;
}

export interface Spike {
  work_item_id: string;
  why: string;
  questions: string[];
  exit_criteria: string;
}

export interface RefinementReadiness {
  level: ReadinessLevel;
  score: number;
  ready_signals: string[];
  needs_clarification: string[];
  investigation_required: string[];
}

export interface RefinementPlan {
  requirement_summary: string;
  objective: string;
  desired_outcome: string;
  scope: string[];
  out_of_scope: string[];
  functional_requirements: string[];
  non_functional_requirements: string[];
  constraints: string[];
  assumptions: string[];
  missing_work_categories: string[];
  work_items: WorkItem[];
  edges: WorkItemEdge[];
  spikes: Spike[];
  open_questions: OpenQuestion[];
  engineering_context_grounded: boolean;
  readiness: RefinementReadiness | null;
  critical_paths: string[][];
  parallelizable_ids: string[];
  unresolved_source_note: string;
  source_jira_key: string | null;
  source_jira_url: string | null;
}

export interface ConversationTurnPayload {
  intent: string;
  resolved_repository_id: string | null;
  resolved_repository_name: string | null;
  why: string;
  evidence: AskEvidenceItem[];
  impact: AskImpact | null;
  actions: AskAction[];
  entities: ConversationEntityRef[];
  /** This turn could not confidently identify which system the question
   * is about, so it asks instead of answering. Carries no evidence and no
   * impact by construction — see `_clarification_turn` on the backend. */
  needs_clarification: boolean;
  candidates: AskRepositoryCandidate[];
  migration: MigrationScope | null;
  refinement: RefinementPlan | null;
  degraded: boolean;
}

export interface ConversationMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  payload: ConversationTurnPayload | null;
  created_at: string;
}

export interface Conversation {
  id: string;
  title: string;
  mode: ConversationMode;
  created_at: string;
  updated_at: string;
  messages: ConversationMessage[];
}

export interface ConversationSummary {
  id: string;
  title: string;
  mode: ConversationMode;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/conversation.py
"""Request/response shapes for `POST /conversations` and
`POST /conversations/{id}/messages` — the Home page's conversational
investigation loop. See `app.services.conversation_service`'s own
docstring for the state model these serialize.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.ask import (
    MAX_QUESTION_LENGTH,
    AskAction,
    AskEvidenceItem,
    AskImpact,
    AskRepositoryCandidate,
)
from app.schemas.migration import MigrationScope
from app.schemas.refinement import RefinementPlan

CONVERSATION_MODES = Literal["general", "migration", "refinement"]


class ConversationEntityRef(BaseModel):
    """One system the investigation has named so far, with a stable
    short-hand (`ref`, e.g. "A") a follow-up can point back at ("what if
    I only fix A and B?") without repeating the real identifier."""

    ref: str
    name: str
    impact_level: Literal["low", "medium", "high"] | None = None


class ConversationTurnPayload(BaseModel):
    """The structured half of an assistant turn — same evidence/impact/
    action vocabulary `AskResponse` already defines (see that schema's
    docstring), plus `entities`, the running reference registry follow-up
    questions resolve against."""

    intent: str
    resolved_repository_id: str | None = None
    resolved_repository_name: str | None = None
    why: str = ""
    evidence: list[AskEvidenceItem] = Field(default_factory=list)
    impact: AskImpact | None = None
    actions: list[AskAction] = Field(default_factory=list)
    entities: list[ConversationEntityRef] = Field(default_factory=list)
    # C-1 — this turn could not confidently identify which system the
    # question is about, so it asks instead of answering. Carries no
    # evidence and no impact by construction; `candidates` is what the
    # user can pick from. The frontend renders a clarification prompt,
    # never the grounded-answer treatment.
    needs_clarification: bool = False
    candidates: list[AskRepositoryCandidate] = Field(default_factory=list)
    # Only set in a "migration" mode conversation, and only once a source
    # technology has actually been grounded — see
    # `app.services.migration_grounding`.
    migration: MigrationScope | None = None
    # Only set in a "refinement" mode conversation — the full current
    # refinement plan (work items, dependency graph, readiness, ...),
    # replaced wholesale each turn rather than diffed — see
    # `app.services.refinement_grounding`.
    refinement: RefinementPlan | None = None
    # Set when the LLM synthesis step itself failed (misconfigured
    # provider, malformed response, ...) — `content` still carries a
    # honest, non-fabricated fallback message in that case; the frontend
    # uses this to skip rendering an "AI Insight" tag on a turn that never
    # actually reasoned about anything.
    degraded: bool = False


class ConversationMessageResponse(BaseModel):
    id: str
    role: Literal["user", "assistant"]
    content: str
    payload: ConversationTurnPayload | None = None
    created_at: datetime


class ConversationResponse(BaseModel):
    id: str
    title: str
    mode: str
    created_at: datetime
    updated_at: datetime
    messages: list[ConversationMessageResponse]


class ConversationSummary(BaseModel):
    id: str
    title: str
    mode: str
    created_at: datetime
    updated_at: datetime


class StartConversationRequest(BaseModel):
    # H-2 — same server-side ceiling as `AskRequest.question`; see
    # `app.schemas.ask.MAX_QUESTION_LENGTH`.
    question: str = Field(max_length=MAX_QUESTION_LENGTH)
    # "migration" opts a conversation into Migration Assistant's grounding/
    # prompt (see `ConversationService`) — same tables, same state model,
    # just a different lens on the same underlying graph. Defaults to
    # "general" so every existing Ask GraphForge caller is unaffected.
    mode: CONVERSATION_MODES = "general"


class PostMessageRequest(BaseModel):
    message: str = Field(max_length=MAX_QUESTION_LENGTH)
